### PR TITLE
Emit deprecation warning if multiple v1 templates match with a new index

### PR DESCRIPTION
**Pull request by probakowski**
This change adds deprecation warning when multiple v1 templates match with a new index. 
This is deprecated because with v2 templates only one template can match and multiple "levels" of settings should be done using component templates.

Relates to #53101

### CODE REVIEW

**Review by dakrone — COMMENTED** [review]
Thanks for changing this @probakowski! Before we merge it could you run a test locally with ESClientYamlTestSuite's method:

```java
@BeforeClass
public static void initializeUseDefaultNumberOfShards() {
    useDefaultNumberOfShards = usually();
}
```

Where `usually()` is changed to `false`? This will install a global template for all the YAML tests, and I want to make sure there are no tests that are going to fail due to warnings (better to fail now before we cause CI failures)

**Comment by probakowski** [comment]
@dakrone that was good idea, there's plenty of failures. I'll work on fixing them and request review again when done

**Comment by probakowski** [comment]
@elasticmachine update branch

**Comment by probakowski** [comment]
@elasticmachine update branch

**Comment by probakowski** [comment]
@elasticmachine run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 1]
```diff
@@ -345,6 +345,12 @@ public ClusterState applyCreateIndexRequest(ClusterState currentState, CreateInd
                 final List<IndexTemplateMetadata> v1Templates = MetadataIndexTemplateService.findV1Templates(currentState.metadata(),
                     request.index(), isHiddenFromRequest);
 
+                if (v1Templates.size() > 1) {
+                    deprecationLogger.deprecatedAndMaybeLog("index_template_multiple_match", "index [{}] matches multiple templates " +
+                        "[{}], and won't be supported with v2 index templates", request.index(),
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java:350]: I think this is a little confusing about what will not be supported and what is not going to be supported going forward, maybe it should be something like:

"index [foo] matches multiple v1 templates [t1, t2], v2 index templates will only match a single index template"

What do you think?

**Review by dakrone — APPROVED** [review]
This LGTM, thanks Przemko! I left a comment about clarifying the warning text though, let me know what you think

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java (hunk 1 above)
Comment (probakowski) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java]: This is good idea, I've changed it

**Comment by probakowski** [comment]
@elasticmachine update branch

**Pull request merged**